### DISCUSSION TITLE
WebRTC / VP8 Support

### DISCUSSION THREAD

**Discussion by mfchiz** [bluenviron/mediamtx#1508] [Questions] [answered]
I've been really impressed with this software. 

I've been able to successfully publish a stream opencv/gstreamer to the server from Unity and view the output in chrome. Using the demo code provided below:

out = cv2.VideoWriter('appsrc ! videoconvert' + \
    ' ! x264enc speed-preset=ultrafast bitrate=600 key-int-max=40' + \
    ' ! rtspclientsink location=rtsp://localhost:8554/mystream',
    cv2.CAP_GSTREAMER, 0, fps, (width, height), True)
    
    
I would like to be able to read the WebRTC video stream in Unity so I'm trying now to convert the signaling process that is in the demo web page to use Unity WebRTC package. I'm running into a problem  that when I connect to the server with the websocket url from Unity, it seems to work, and I get a message back with the STUN Servers, but then when I try to send an "Offer" message to the server over the wqebsocket it closes the connection immediately and my client code gets a . Here is the output from the log:
  
2023/02/24 12:20:51 INF [WebRTC] [conn 127.0.0.1:50478] opened
2023/02/24 12:20:52 DEB [WebRTC] [conn 127.0.0.1:50478] peer connection state: closed
2023/02/24 12:20:52 INF [WebRTC] [conn 127.0.0.1:50478] closed (unable to start track, codec is not supported by remote)

I think this could be because I am using Unity on Linux and there is no h264 encoder support in the unity player - only VP8/VP9.  I am however able to publish with h264 with gstreamer.  
I see the server supports WebRTC and VP8. Does the endpoint transcode too VP8  by default or do I need to update the gstreamer code that is publishing? 

I would love any thoughts. Thanks!

**Comment by mfchiz** [comment, marked as answer]
Hi I thought i'd follow up because I found a couple solutions to my problem. If the rtsp ingest is in h264 then the WebRTC endpoint provides h264. The WebRTC endpoints do not magically transcode to other codecs by default - which  makes sense. To have the WebRTC endpoint provide VP8 or VP9 you need either ingest with the stream in that codec or setup the server to transcode. Here is a simple way to ingest VP9 like below.

  	writer.open("appsrc ! videoconvert ! vp9enc cpu-used=5 ! queue  "+
						"! rtspclientsink location=rtsp://localhost:8554/vp9",
   					Videoio.CAP_GSTREAMER, 0,  30, new Size(width, height), true);

This works pretty well with about a second delay in the content. The second way is to transcode stream after it has been ingested  using the instructions provided. Here is how i did that: 

    runOnReady: ffmpeg -i rtsp://localhost:$RTSP_PORT/$RTSP_PATH -pix_fmt yuv420p -c:v libvpx -qmin 0 -qmax 50  -f rtsp rtsp://localhost:$RTSP_PORT/compressed
    runOnReadyRestart: yes

This was much slower but I think it is related to the parameters passed to FFMpeg. 

I hope this helps someone. 

Also fantastic work with this software it has been really easy to use so far. Thank you!

**Comment by aler9** [comment]
Hello,

> The WebRTC endpoints do not magically transcode to other codecs by default - which makes sense.

This server tries in any possible way to avoid re-encoding, since re-encoding requires a lot of resources and makes it impossible to scale the server on most machines. The server only performs remuxing, i.e. it changes the outer layer of packets, for instance by unwrapping H264 packets from RTSP and wrapping them into WebRTC. If you need a specific codec, my advice is to tune the source of the stream like you did.

[a comment by an automated account is not shown]
